### ISSUE TITLE
LaunchedClassLoader issue Illegal character in opaque part when log4j2.xml (or any file) is in a jar file 

### ISSUE THREAD

**Issue by funnelspider** [spring-projects/spring-boot#38827]
I have a spring boot application which is started via 

```
java -jar app.jar 
```

It uses log4j2 as a logging framework

Recently I updated to java 21 (from java 17) and to spring boot 3.1.5 / 3.2.0  from spring boot 2.7.4

With the old java and spring boot 2.7.4 log4j find the log4j2.xml with no issues 

```
DEBUG StatusLogger Apache Log4j Core 2.17.2 initializing configuration XmlConfiguration[location=jar:file:/C:/Program%20Files%20(x86)/my_app/app.jar!/BOOT-INF/classes!/log4j2.xml]
```

With the new spring boot 3.1.5 it fails 

```
WARN StatusLogger Error accessing jar:nested:/C:/Program Files (x86)/my_app/app.jar/!BOOT-INF/classes/!/log4j2.xml due to Illegal character in opaque part at index 22: jar:nested:/C:/Program Files (x86)/my_app/app.jar/!BOOT-INF/classes/!/log4j2.xml, ignoring.
```

Far as I can tell the old spring boot was using the `LaunchedUrlClassLoader` while the new one is using `LaunchedClassLoader`
 
This in turn uses a wrong format for the URL (spaces not converted to %20?)

**Comment by philwebb** [comment]
This may well be a duplicate of #38660. @funnelspider are you able to give the latest `3.2.1-SNAPSHOT` a try to see if we've already fixed this one?

**Comment by funnelspider** [comment]
tried it right now and with version 3.2.1-SNAPSHOT i still get the same error 

![image](https://github.com/spring-projects/spring-boot/assets/59641306/5c70c589-f0a7-49bd-b591-f6059ae247a1)

**Comment by funnelspider** [comment]
I couldn't reproduce the issue with a new spring boot project so I started checking what is wrong in my project
It seems that the spring-boot-maven-plugin wasn't using the 3.2.1-SNAPSHOT version and I was getting the wrong spring-boot loader. 

Forcing the correct 3.2.1-SNAPSHOT plugin seems to do the trick

  <plugin>
                <groupId>org.springframework.boot</groupId>
                <artifactId>spring-boot-maven-plugin</artifactId>
				<version>3.2.1-SNAPSHOT</version>
            	<executions>
	                <execution>
	                    <goals>
	                        <goal>repackage</goal>
	                    </goals>
	                </execution>
	            </executions>
            </plugin>

So it is indeed fixed in the 3.2.1-SNAPSHOT build. 

When will this be released ?

**Comment by scottfrederick** [comment]
Thanks for trying the snapshot and letting us know it fixed your problem. 

You can see the scheduled dates for releases on the GitHub [milestones page](https://github.com/spring-projects/spring-boot/milestones) and on the [Spring release calendar](https://calendar.spring.io/).
